File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool ans ;
    public bool IsBalanced(TreeNode root) {
        //abs(r-l ) < 1

        ans = true;
        func(root);
        return ans;



    }
    public int func(TreeNode root){
        if (root is null) return 0;

        int left = func(root.left);
        int right = func(root.right);

        if (Math.Abs(right - left ) > 1){
            ans = false;
        }

        return 1 + Math.Max(left, right);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        #edge case 
        if not root:
            return 0
        #get the max depth from both sides

        #make global to class
        self.deepest = 0 

        self.dfs(root)
        return self.deepest
    
    def dfs(self,root):
        if root is None:
            return 0
        
        #get values
        left = self.dfs(root.left)
        right = self.dfs(root.right)

        self.deepest = max(self.deepest, left + right)

        return 1 + max(left,right)